### PR TITLE
Snapshot selected globals and restore them in spawned process

### LINKED ISSUE #12685: Deterministic flags and other globals do not get transferred to spawned processes
## 🐛 Bug

When using the ddp-spawn strategy, global flags like the torch.use_deterministic_algorithms flags set by the Trainer do not get transferred to the worker processes. The reason is that torch gets re-imported when spawning the workers (naturally) and thus the globals are back to defaults.  

### To Reproduce

```python
import os

import torch
from torch.utils.data import DataLoader, Dataset

from pytorch_lightning import LightningModule, Trainer


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        assert torch.are_deterministic_algorithms_enabled()
        loss = self(batch).sum()
        return {"loss": loss}

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)


def run():
    train_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    model = BoringModel()
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        limit_train_batches=1,
        limit_val_batches=1,
        limit_test_batches=1,
        num_sanity_val_steps=0,
        max_epochs=1,
        enable_model_summary=False,
        enable_progress_bar=False,
        strategy="ddp_spawn",
        devices=2,
        accelerator="cpu",
        deterministic=True,
    )
    trainer.fit(model, train_dataloaders=train_data)


if __name__ == "__main__":
    run()

```

### Expected behavior

See assertion in the code. 

### Environment

PyTorch 1.10 and 1.11
PL 1.6 (master)

### Additional context

I dreamed about it last night (no kidding) and checked this morning and the bug turned out to be real.

This applies to all globals from torch, not just deterministic flags. We already handle this problem for the seed in the `DDPSpawnStrategy._worker_setup()` method.

cc @tchaton @rohitgr7 @justusschock @kaushikb11 @awaelchli @akihironitta

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #12685

The problem: Globals like torch deterministic flags or the seed get lost when spawning processes. This is caused because essentially the torch module gets re-imported in the child process, leading to globals being re-initialized. This is simply the nature of how `mp.spawn` works. This means it only affects ddp spawn (multiprocessing start method "spawn"). 

- Subprocess launch is not affected because all processes start from an identical state by design
- Fork is not affected because all children inherit the exact memory of the parent process

We can solve this by picking the globals we care about, save them, send them to the child process and restore them there before we launch the trainer run methods. Hence, this PR implements this directly in the launcher. 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

cc @borda @justusschock @kaushikb11 @awaelchli @akihironitta @rohitgr7

### CODE REVIEW

File: src/pytorch_lightning/strategies/ddp_spawn.py [hunk 1]
```diff
@@ -175,7 +174,6 @@ def set_world_ranks(self, process_idx: int = 0) -> None:
         rank_zero_only.rank = self.cluster_environment.global_rank()
 
     def _worker_setup(self, process_idx: int) -> None:
-        reset_seed()
```
Comment (awaelchli) [inline src/pytorch_lightning/strategies/ddp_spawn.py:178]: This is now handled in the launcher, so not necessary anymore. Would anyway not have been necessary when using fork.

**Review by akihironitta — APPROVED** [review]
LGTM!

**Review by carmocca — APPROVED** [review]

**Review by Borda — APPROVED** [review]

File: src/pytorch_lightning/strategies/launchers/multiprocessing.py [hunk 2]
```diff
@@ -209,3 +223,53 @@ class _WorkerOutput(NamedTuple):
...
+
+            # in main process
+            snapshot = _GlobalStateSnapshot.capture()
+
+            # in worker process
+            snapshot.restore()
+    """
+
+    use_deterministic_algorithms: bool
+    use_deterministic_algorithms_warn_only: bool
```
Comment (rohitgr7) [inline src/pytorch_lightning/strategies/launchers/multiprocessing.py:247]: maybe capture these 2 in a dict? then you can pass it by checking the `torch_version` only once.

**Review by rohitgr7 — APPROVED** [review]

File: src/pytorch_lightning/strategies/launchers/multiprocessing.py [hunk 3]
```diff
@@ -209,3 +223,53 @@ class _WorkerOutput(NamedTuple):
...
+        that we want to restore in a spawned worker process.
+        """
+        return cls(
+            use_deterministic_algorithms=torch.are_deterministic_algorithms_enabled(),
+            use_deterministic_algorithms_warn_only=torch.is_deterministic_algorithms_warn_only_enabled()
+            if _TORCH_GREATER_EQUAL_1_11
+            else False,
+            cudnn_benchmark=torch.backends.cudnn.benchmark,
+            rng_states=_collect_rng_states(),
+        )
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply src/pytorch_lightning/strategies/launchers/multiprocessing.py]: done

File: src/pytorch_lightning/strategies/launchers/multiprocessing.py (hunk 2 above)
Comment (awaelchli) [reply src/pytorch_lightning/strategies/launchers/multiprocessing.py:247]: Tried but makes it only more complicated, and requires to introduce new names. I prefer the explicit way as it is also easy to annotate with types.
The torch version check can be removed once our minimal version is 1.11.

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by kaushikb11 — APPROVED** [review]
